### PR TITLE
feat: InsertionPageType on prepareForLogging\rTESTING=unit

### PR DESCRIPTION
**Pull request by scottmcmaster**
@danbosnichill would you mind double-checking the logic and (especially) the names of things before I add tests? Thanks!

### CODE REVIEW

**Review by danbosnichill — COMMENTED** [review]
The logic looks good to me.

**Comment by scottmcmaster** [comment]
> The logic looks good to me.

Thanks, I added a couple of tests and took off the "WIP", feel free to have a final look.

File: src/client.ts [hunk 1]
```diff
@@ -165,6 +165,10 @@ export class PromotedClientImpl implements PromotedClient {
       if (error) {
         this.handleError(error);
       }
+
+      if (this.shadowTrafficDeliveryPercent > 0 && metricsRequest.insertionPageType != InsertionPageType.Unpaged) {
```
Comment (danbosnichill) [inline src/client.ts:169]: Depending on which paging value is set, it'd impact:
1. the offset used when calculating `position`.
2. which page to take from unpaged when trimming to a page.

**Review by danbosnichill — CHANGES_REQUESTED** [review]

File: src/client.ts (hunk 1 above)
Comment (scottmcmaster) [reply src/client.ts:169]: We use the paging parameters for that currently. Are you saying that we'd override that somehow, and/or change the behavior when it's not provided (and how?)?

That's somewhat orthogonal to shadow traffic, though, right? At any rate it's in a different part of the code.

File: README.md [hunk 2]
```diff
@@ -115,11 +115,15 @@ There are two ways of doing this with `PromotedClient`:
 1. You can use `deliver` but add a `shouldOptimize: false` property.
 2. You can use `prepareForLogging` method call instead. The `prepareForLogging` signature is similar to `deliver` and should be integrated the same way.
 
-## Enabled setting
+## Pagination
```
Comment (scottmcmaster) [inline README.md:118]: @danbosnichill please review, thanks!

File: src/client.ts (hunk 1 above)
Comment (scottmcmaster) [reply src/client.ts:187]: I modified the logic in applyPaging per our discuss, please double-check.

File: README.md [hunk 3]
```diff
@@ -115,11 +115,15 @@ There are two ways of doing this with `PromotedClient`:
...
+- When calling `deliver`, we expect that you will pass an unpaged (complete) list of insertions, and the SDK assumes this to be the case.
 
-## Pagination
+- When calling `prepareForLogging` with shadow traffic turned on, we also expect an unpaged list of insertions, since in this case we are simulating delivery.
+
+- When calling `prepareForLogging` otherwise, you may choose to pass "pre-paged" or "unpaged" insertions based on the `insertionPageType` field on the `MetricsRequest`.
+  - When `insertionPageType` is "unpaged", the `Request.paging.offset` and `Request.paging.size` parameters are used to log a "window" of insertions, and the `position` of the insertions will be offset accordingly.
+  - When `insertionPageType` is "pre-paged", the SDK assumes the client is handling paging, and the position offset is assumed to be 0.
```
Comment (danbosnichill) [inline README.md:126]: Maybe rephrase this case.  `offset` should still be set.  The SDK won't paginate.

File: src/client.ts [hunk 4]
```diff
@@ -216,7 +221,17 @@ export class PromotedClientImpl implements PromotedClient {
       if (error) {
         this.handleError(error);
       }
+
+      // It's ok to not specify insertionPageType because for delivery we assume it is
+      // "unpaged"...but if you set it, it had best not be "prepaged".
```
Comment (danbosnichill) [inline src/client.ts:226]: As we talked about in person, I like your recommendation about breaking down the input counts so this is harder to screw up.  If this isn't feasible, we might want to force the parameter on `deliver` too (to lower the chance of errors).

File: src/client.ts [hunk 5]
```diff
@@ -371,9 +387,18 @@ export class PromotedClientImpl implements PromotedClient {
...
    */
-  private applyPaging = (insertions: Insertion[], paging?: Paging): Insertion[] => {
+  private applyPaging = (
+    insertions: Insertion[],
+    paging?: Paging,
+    insertionPageType?: InsertionPageType
+  ): Insertion[] => {
     const insertionPage: Insertion[] = [];
     let start = paging?.offset ?? 0;
+    if (insertionPageType === InsertionPageType.PrePaged) {
```
Comment (danbosnichill) [inline src/client.ts:397]: I think lines 407-414 are messed up.   If it's unpaged and we're running delivery in this client, we'll to do an offset into the list of insertions before producing a page.

**Review by danbosnichill — CHANGES_REQUESTED** [review]

File: src/client.ts (hunk 4 above)
Comment (scottmcmaster) [reply src/client.ts:226]: The counts breakdown, we might want to make sure we've covered all the cases and review the proposal, then follow up on that in a subsequent PR.

For "forcing" in the current approach, are you talking about changing this code so that it treats "unset" as an error case rather than assuming unpaged, and/or going beyond the current "call handleError if performChecks" all the way to "raise an exception and stop the flow"?

File: src/client.ts (hunk 5 above)
Comment (scottmcmaster) [reply src/client.ts]: Done in paging.ts.

**Review by scottmcmaster — COMMENTED** [review]
After some refactoring and our discussions, I think the approach as realized in code is more principled now, please double-check one more time, thanks :-).

File: src/pager.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+   * @param assignPosition whether or not to assign a position to the insertions, which deliver requests do and log requests may not.
+   * @param paging paging info, may be nil
+   * @param insertionPageType the type of paging the client wants, defaults to "unpaged"
+   * @returns the modified page of insertions
+   */
+  applyPaging = (
+    insertions: Insertion[],
+    assignPosition: boolean,
+    paging?: Paging,
+    insertionPageType?: InsertionPageType
```
Comment (danbosnichill) [inline src/pager.ts:18]: Might be better to force the SDK-internal methods to set it.

File: src/pager.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    for (let index = start; index < insertions.length; index++) {
+      if (insertionPage.length >= size) {
+        break;
+      }
+
+      const insertion = insertions[index];
+      if (assignPosition) {
+        insertion.position = start;
```
Comment (danbosnichill) [inline src/pager.ts:39]: Something seems off.  It looks like index and start are equal in this loop.  

I think the PrePaged case is broken.  By be easier to read if you rename 'start' to something like 'startOffset'.

```
    let offset = paging?.offset ?? 0;
    let index = offset;
    if (insertionPageType === InsertionPageType.PrePaged) {
      // When insertions are pre-paged, we ignore any provided offset.
      index = 0;
    }

    let size = paging?.size ?? -1;
    if (size <= 0) {
      size = insertions.length;
    }

    const insertionPage: Insertion[] = [];
    while (index < insertions.length && insertionPage.length < size) {
      const insertion = insertions[index];
      if (assignPosition) {
        insertion.position = offset;
      }
      insertionPage.push(insertion);
      index++;
      offset++;
    }

    return insertionPage;
```

File: src/pager.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  /**
+   * Sets the correct position field on each assertion based on paging parameters and takes
+   * a page of full insertions (if necessary).
+   * @param insertions the full set of insertions
+   * @param assignPosition whether or not to assign a position to the insertions, which deliver requests do and log requests may not.
+   * @param paging paging info, may be nil
+   * @param insertionPageType the type of paging the client wants, defaults to "unpaged"
+   * @returns the modified page of insertions
+   */
+  applyPaging = (
```
Comment (danbosnichill) [inline src/pager.ts:14]: Might make sense to have a unit test just for applyPaging.

File: src/pager.ts (hunk 7 above)
Comment (scottmcmaster) [reply src/pager.ts]: It worked like I intended, which means I'm still confused about what we are trying to accomplish with position assignment during paging :-(. For now I will just change to your code.

File: src/client.ts [hunk 9]
```diff
@@ -193,7 +199,7 @@ export class PromotedClientImpl implements PromotedClient {
 
     // If defined, log the Request to Metrics API.
     const { fullInsertion, request } = metricsRequest;
-    const insertion = !request?.paging?.size ? fullInsertion : fullInsertion.slice(0, request.paging.size);
+    const insertion = new Pager().applyPaging(fullInsertion, false, request?.paging, metricsRequest.insertionPageType);
```
[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.ts]: > This is cleaner.
> 
> Not for this PR - I think splitting and naming the different types of insertions (fullInsertion, requestInsertion, responseInsertion) differently might help.

I think if I were to try to do that in this PR, it would mess everything up, so I'd prefer to defer that.

[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.ts]: > Please add a comment for when assignPosition should be false vs true.
> 
> I'm a little worried we're now not assigning position on the responseInsertions that are actually logged. It looks like we aren't when skimming the code.
> 
> I think it's fine to move the responseInsertion applyPaging logic to only happen inside prepareForLogging/deliver (and not in the deferred createLogRequestFn). I originally did that as an optimization.

I think it's this comment that's already there (?): "which deliver requests do and log requests may not"

[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.ts]: > Please add a comment for when assignPosition should be false vs true.
> 
> I'm a little worried we're now not assigning position on the responseInsertions that are actually logged. It looks like we aren't when skimming the code.
> 
> I think it's fine to move the responseInsertion applyPaging logic to only happen inside prepareForLogging/deliver (and not in the deferred createLogRequestFn). I originally did that as an optimization.

It seems difficult to do that without a bigger redo of the structure, since the paged insertions need to go on the log request, but the log request doesn't exist until the deferred function gets called.

I thought we didn't want position on log-only requests for, I'm not sure why actually. What do you think about changing applyPaging to assign position if and only if it's not already set? I think that would solve a lot of confusion.

File: src/pager.ts (hunk 8 above)
Comment (scottmcmaster) [reply src/pager.ts:14]: I will wait for replies to some of the other questions first, but yes.

File: src/client.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (danbosnichill) [reply src/client.ts]: > I think if I were to try to do that in this PR, it would mess everything up, so I'd prefer to defer that.

Yea.  Not for this PR.

> but the log request doesn't exist until the deferred function gets called.

It's fine to change this.

> What do you think about changing applyPaging to assign position if and only if it's not already set? I think that would solve a lot of confusion.

I'm open to that if we only applyPaging on responseInsertions.

File: src/pager.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import { InsertionPageType } from './insertion-page-type';
+import { Insertion, Paging } from './types/delivery';
+
+export class Pager {
+  /**
+   * Sets the correct position field on each assertion based on paging parameters and takes
+   * a page of full insertions (if necessary).
+   * @param insertions the full set of insertions
```
[the remark this answers is not in the document]
Comment (danbosnichill) [reply src/pager.ts:8]: `LogRequest.insertion` are responseInsertions.

In the upcoming synthetic DeliveryLog work, I'll stick them on `DeliveryLog.response.insertion`.

File: src/pager.ts (hunk 6 above)
Comment (danbosnichill) [reply src/pager.ts:18]: `insertionPageType: InsertionPageType` then make the callers convert from undefined.

Comment (scottmcmaster) [reply src/pager.ts]: Ok, it's a required field here as well as (necessarily) on the request types now.

File: src/client.test.ts [hunk 11]
```diff
@@ -1467,17 +1556,14 @@ describe('metrics', () => {
         toInsertion(newProduct('3'), {
           insertionId: 'uuid1',
           requestId: 'uuid0',
```
[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.test.ts:1582]: My tentative change to always assign position if undefined in applyPaging would fix this.

File: src/pager.ts [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    if (size <= 0) {
+      size = insertions.length;
+    }
+    const insertionPage: Insertion[] = [];
+    while (index < insertions.length && insertionPage.length < size) {
+      const insertion = insertions[index];
+      if (insertion.position === undefined) {
+        insertion.position = offset;
+      }
+      insertionPage.push(insertion);
```
Comment (danbosnichill) [inline src/pager.ts:37]: perf nit - when we do perf optimization, we should avoid push and set the array size directly.

This perf test isn't accurate since it runs locally.
https://jsbench.me/rbkrblk7jr/1

**Review by danbosnichill — APPROVED** [review]

File: src/client.ts (hunk 4 above)
Comment (scottmcmaster) [reply src/client.ts]: The field/parameter is required now on the request types.

File: src/pager.ts (hunk 12 above)
Comment (scottmcmaster) [reply src/pager.ts]: I just did it.

File: src/client.test.ts [hunk 13]
```diff
@@ -219,14 +241,73 @@ describe('no-op', () => {
           },
         },
         fullInsertion: toInsertions(products),
+        insertionPageType: InsertionPageType.Unpaged,
       });
-      expect(response.insertion).toEqual(toInsertions([newProduct('3')]));
+
+      // Paging parameters advance to the second insertion.
+      const expectedInsertions = toResponseInsertions([newProduct('2')]);
```
[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.test.ts:244]: In this case, I think the extra line emphasizes the purpose of the test.

File: src/client.test.ts [hunk 14]
```diff
@@ -340,18 +408,22 @@ describe('deliver', () => {
       expect(deliveryClient.mock.calls.length).toBe(0);
       expect(metricsClient.mock.calls.length).toBe(0);
 
+      // SDK-provided positions
       expect(response.insertion).toEqual([
         toInsertion(newProduct('3'), {
           insertionId: 'uuid1',
           requestId: 'uuid0',
+          position: 0,
```
[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.test.ts:427]: I think this change is correct -- it's SDK-delivery-provided positions. Maybe clarify why you weren't expecting that here in that case because I might be missing something.

File: src/client.test.ts [hunk 15]
```diff
@@ -614,6 +689,7 @@ describe('deliver', () => {
 
     it('toCompactMetricsInsertion arm=CONTROL defaultRequestValues', async () => {
       const deliveryClient: any = jest.fn(failFunction('Delivery should not be called in CONTROL'));
+      // Should not have position set due to the custom compact function we're testing with.
```
[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.test.ts:706]: This is a case I mentioned yesterday: Because compaction runs post-apply-paging, if the SDK client gives us a compaction function that strips position, we don't have any control over it...(?)

File: src/client.test.ts [hunk 16]
```diff
@@ -545,15 +639,9 @@ describe('deliver', () => {
...
-          toInsertionOnlyContentId(newProduct('3'), {
-            position: 0,
-          }),
-          toInsertionOnlyContentId(newProduct('2'), {
-            position: 1,
-          }),
-          toInsertionOnlyContentId(newProduct('1'), {
-            position: 2,
-          }),
+          toInsertionOnlyContentId(newProduct('3')),
```
[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.test.ts:638]: This has to do with the reply to a comment I made below.

File: src/client.test.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (danbosnichill) [reply src/client.test.ts:706]: Ah, yea.  Currently, if they have a bad compact function, we break.

One idea is to change compact to only impact properties.

File: src/client.test.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (danbosnichill) [reply src/client.test.ts:427]: Yes, this is correct.  I'm just complaining about GitHub.

File: src/client.test.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (danbosnichill) [reply src/client.test.ts:638]: Got it.  Yea, we don't want clients to be able to remove these fields.  The test is correct but the client call is bad.

Longer-term, we should probably change compact to just impact properties.

**Review by danbosnichill — APPROVED** [review]
I only skimmed the updates.  GitHub does a horrible job showing diffs.  It isn't comparing against the last commit I reviewed.  :-(

File: src/pager.ts (hunk 8 above)
Comment (scottmcmaster) [reply src/pager.ts]: Did this.

File: src/client.test.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.test.ts:638]: Per our discussion, I opened an issue for this in the backlog.

File: src/client.test.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (scottmcmaster) [reply src/client.test.ts:706]: Per our discussion, issue opened in the backlog.
